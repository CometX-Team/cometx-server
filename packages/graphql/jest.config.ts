module.exports = {
  displayName: 'graphql',
  preset: '../../jest.preset.ts',
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
    },
  },
  testEnvironment: 'node',
  transform: {
    '^.+\\.[tj]s$': 'ts-jest',
  },
  moduleFileExtensions: ['ts', 'js', 'html'],
  coverageDirectory: '../../coverage/packages/graphql',
};
