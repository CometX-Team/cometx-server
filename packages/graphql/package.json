{
  "name": "@cometx-server/graphql",
  "version": "0.0.1",
  "type": "commonjs"
}
