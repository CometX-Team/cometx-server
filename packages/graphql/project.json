{
  "root": "packages/graphql",
  "sourceRoot": "packages/graphql/src",
  "targets": {
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/graphql",
        "tsConfig": "packages/graphql/tsconfig.lib.json",
        "packageJson": "packages/graphql/package.json",
        "main": "packages/graphql/src/index.ts",
        "assets": ["packages/graphql/*.md"]
      }
    },
    "publish": {
      "executor": "@nrwl/workspace:run-commands",
      "options": {
        "command": "node tools/scripts/publish.mjs graphql {args.ver} {args.tag}",
        "cwd": "dist/packages/graphql"
      },
      "dependsOn": [
        {
          "projects": "self",
          "target": "build"
        }
      ]
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/graphql/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/packages/graphql"],
      "options": {
        "jestConfig": "packages/graphql/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
