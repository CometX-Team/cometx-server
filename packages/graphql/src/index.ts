export * from './lib/create-graphql-options';
export * from './lib/graphql-custom-fields';
