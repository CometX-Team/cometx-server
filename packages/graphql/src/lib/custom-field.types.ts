import { CustomFieldType, Injector, Type } from '@cometx-server/common';
import { CometXEntity } from '@cometx-server/entity';
import {
  BooleanCustomFieldConfig as GraphQLBooleanCustomFieldConfig,
  CustomField,
  DateTimeCustomFieldConfig as GraphQLDateTimeCustomFieldConfig,
  FloatCustomFieldConfig as GraphQLFloatCustomFieldConfig,
  IntCustomFieldConfig as GraphQLIntCustomFieldConfig,
  LocaleStringCustomFieldConfig as GraphQLLocaleStringCustomFieldConfig,
  LocalizedString,
  RelationCustomFieldConfig as GraphQLRelationCustomFieldConfig,
  StringCustomFieldConfig as GraphQLStringCustomFieldConfig,
  TextCustomFieldConfig as GraphQLTextCustomFieldConfig,
} from './generated-custom-field.types';

export type DefaultValueType<T extends CustomFieldType> = T extends
  | 'string'
  | 'localeString'
  ? string
  : T extends 'int' | 'float'
  ? number
  : T extends 'boolean'
  ? boolean
  : T extends 'datetime'
  ? Date
  : T extends 'relation'
  ? any
  : never;

export type BaseTypedCustomFieldConfig<
  T extends CustomFieldType,
  C extends CustomField,
> = Omit<C, '__typename' | 'list'> & {
  type: T;
  /**
   * @description
   * Whether or not the custom field is available via the Shop API.
   * @default true
   */
  public?: boolean;
  nullable?: boolean;
  unique?: boolean;
};

/**
 * @description
 * Configures a custom field on an entity in the {@link CustomFields} config object.
 *
 * @docsCategory custom-fields
 */
export type TypedCustomSingleFieldConfig<
  T extends CustomFieldType,
  C extends CustomField,
> = BaseTypedCustomFieldConfig<T, C> & {
  list?: false;
  defaultValue?: DefaultValueType<T>;
  validate?: (
    value: DefaultValueType<T>,
    injector: Injector,
  ) =>
    | string
    | LocalizedString[]
    | void
    | Promise<string | LocalizedString[] | void>;
};

export type TypedCustomListFieldConfig<
  T extends CustomFieldType,
  C extends CustomField,
> = BaseTypedCustomFieldConfig<T, C> & {
  list?: true;
  defaultValue?: Array<DefaultValueType<T>>;
  validate?: (
    value: Array<DefaultValueType<T>>,
  ) => string | LocalizedString[] | void;
};

export type TypedCustomFieldConfig<
  T extends CustomFieldType,
  C extends CustomField,
> = BaseTypedCustomFieldConfig<T, C> &
  (TypedCustomSingleFieldConfig<T, C> | TypedCustomListFieldConfig<T, C>);

export type StringCustomFieldConfig = TypedCustomFieldConfig<
  'string',
  GraphQLStringCustomFieldConfig
>;
export type LocaleStringCustomFieldConfig = TypedCustomFieldConfig<
  'localeString',
  GraphQLLocaleStringCustomFieldConfig
>;
export type TextCustomFieldConfig = TypedCustomFieldConfig<
  'text',
  GraphQLTextCustomFieldConfig
>;
export type IntCustomFieldConfig = TypedCustomFieldConfig<
  'int',
  GraphQLIntCustomFieldConfig
>;
export type FloatCustomFieldConfig = TypedCustomFieldConfig<
  'float',
  GraphQLFloatCustomFieldConfig
>;
export type BooleanCustomFieldConfig = TypedCustomFieldConfig<
  'boolean',
  GraphQLBooleanCustomFieldConfig
>;
export type DateTimeCustomFieldConfig = TypedCustomFieldConfig<
  'datetime',
  GraphQLDateTimeCustomFieldConfig
>;
export type RelationCustomFieldConfig = TypedCustomFieldConfig<
  'relation',
  Omit<GraphQLRelationCustomFieldConfig, 'entity' | 'scalarFields'>
> & { entity: Type<CometXEntity>; graphQLType?: string; eager?: boolean };

/**
 * @description
 * An object used to configure a custom field.
 *
 * @docsCategory custom-fields
 */
export type CustomFieldConfig =
  | StringCustomFieldConfig
  | LocaleStringCustomFieldConfig
  | TextCustomFieldConfig
  | IntCustomFieldConfig
  | FloatCustomFieldConfig
  | BooleanCustomFieldConfig
  | DateTimeCustomFieldConfig
  | RelationCustomFieldConfig;
