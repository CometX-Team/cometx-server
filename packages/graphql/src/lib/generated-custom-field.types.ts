export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  JSON: any;
};

export type BooleanCustomFieldConfig = CustomField & {
  __typename?: 'BooleanCustomFieldConfig';
  description?: Maybe<Array<LocalizedString>>;
  internal?: Maybe<Scalars['Boolean']>;
  label?: Maybe<Array<LocalizedString>>;
  list: Scalars['Boolean'];
  name: Scalars['String'];
  nullable?: Maybe<Scalars['Boolean']>;
  readonly?: Maybe<Scalars['Boolean']>;
  type: Scalars['String'];
  ui?: Maybe<Scalars['JSON']>;
};

export type CustomField = {
  description?: Maybe<Array<LocalizedString>>;
  internal?: Maybe<Scalars['Boolean']>;
  label?: Maybe<Array<LocalizedString>>;
  list: Scalars['Boolean'];
  name: Scalars['String'];
  nullable?: Maybe<Scalars['Boolean']>;
  readonly?: Maybe<Scalars['Boolean']>;
  type: Scalars['String'];
  ui?: Maybe<Scalars['JSON']>;
};

export type CustomFieldConfig = BooleanCustomFieldConfig | DateTimeCustomFieldConfig | FloatCustomFieldConfig | IntCustomFieldConfig | LocaleStringCustomFieldConfig | RelationCustomFieldConfig | StringCustomFieldConfig | TextCustomFieldConfig;

/**
 * Expects the same validation formats as the `<input type="datetime-local">` HTML element.
 * See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/datetime-local#Additional_attributes
 */
export type DateTimeCustomFieldConfig = CustomField & {
  __typename?: 'DateTimeCustomFieldConfig';
  description?: Maybe<Array<LocalizedString>>;
  internal?: Maybe<Scalars['Boolean']>;
  label?: Maybe<Array<LocalizedString>>;
  list: Scalars['Boolean'];
  max?: Maybe<Scalars['String']>;
  min?: Maybe<Scalars['String']>;
  name: Scalars['String'];
  nullable?: Maybe<Scalars['Boolean']>;
  readonly?: Maybe<Scalars['Boolean']>;
  step?: Maybe<Scalars['Int']>;
  type: Scalars['String'];
  ui?: Maybe<Scalars['JSON']>;
};

export type FloatCustomFieldConfig = CustomField & {
  __typename?: 'FloatCustomFieldConfig';
  description?: Maybe<Array<LocalizedString>>;
  internal?: Maybe<Scalars['Boolean']>;
  label?: Maybe<Array<LocalizedString>>;
  list: Scalars['Boolean'];
  max?: Maybe<Scalars['Float']>;
  min?: Maybe<Scalars['Float']>;
  name: Scalars['String'];
  nullable?: Maybe<Scalars['Boolean']>;
  readonly?: Maybe<Scalars['Boolean']>;
  step?: Maybe<Scalars['Float']>;
  type: Scalars['String'];
  ui?: Maybe<Scalars['JSON']>;
};

export type IntCustomFieldConfig = CustomField & {
  __typename?: 'IntCustomFieldConfig';
  description?: Maybe<Array<LocalizedString>>;
  internal?: Maybe<Scalars['Boolean']>;
  label?: Maybe<Array<LocalizedString>>;
  list: Scalars['Boolean'];
  max?: Maybe<Scalars['Int']>;
  min?: Maybe<Scalars['Int']>;
  name: Scalars['String'];
  nullable?: Maybe<Scalars['Boolean']>;
  readonly?: Maybe<Scalars['Boolean']>;
  step?: Maybe<Scalars['Int']>;
  type: Scalars['String'];
  ui?: Maybe<Scalars['JSON']>;
};

export type LocaleStringCustomFieldConfig = CustomField & {
  __typename?: 'LocaleStringCustomFieldConfig';
  description?: Maybe<Array<LocalizedString>>;
  internal?: Maybe<Scalars['Boolean']>;
  label?: Maybe<Array<LocalizedString>>;
  length?: Maybe<Scalars['Int']>;
  list: Scalars['Boolean'];
  name: Scalars['String'];
  nullable?: Maybe<Scalars['Boolean']>;
  pattern?: Maybe<Scalars['String']>;
  readonly?: Maybe<Scalars['Boolean']>;
  type: Scalars['String'];
  ui?: Maybe<Scalars['JSON']>;
};

export type LocalizedString = {
  __typename?: 'LocalizedString';
  value: Scalars['String'];
};

export type RelationCustomFieldConfig = CustomField & {
  __typename?: 'RelationCustomFieldConfig';
  description?: Maybe<Array<LocalizedString>>;
  entity: Scalars['String'];
  internal?: Maybe<Scalars['Boolean']>;
  label?: Maybe<Array<LocalizedString>>;
  list: Scalars['Boolean'];
  name: Scalars['String'];
  nullable?: Maybe<Scalars['Boolean']>;
  readonly?: Maybe<Scalars['Boolean']>;
  scalarFields: Array<Scalars['String']>;
  type: Scalars['String'];
  ui?: Maybe<Scalars['JSON']>;
};

export type StringCustomFieldConfig = CustomField & {
  __typename?: 'StringCustomFieldConfig';
  description?: Maybe<Array<LocalizedString>>;
  internal?: Maybe<Scalars['Boolean']>;
  label?: Maybe<Array<LocalizedString>>;
  length?: Maybe<Scalars['Int']>;
  list: Scalars['Boolean'];
  name: Scalars['String'];
  nullable?: Maybe<Scalars['Boolean']>;
  options?: Maybe<Array<StringFieldOption>>;
  pattern?: Maybe<Scalars['String']>;
  readonly?: Maybe<Scalars['Boolean']>;
  type: Scalars['String'];
  ui?: Maybe<Scalars['JSON']>;
};

export type StringFieldOption = {
  __typename?: 'StringFieldOption';
  label?: Maybe<Array<LocalizedString>>;
  value: Scalars['String'];
};

export type TextCustomFieldConfig = CustomField & {
  __typename?: 'TextCustomFieldConfig';
  description?: Maybe<Array<LocalizedString>>;
  internal?: Maybe<Scalars['Boolean']>;
  label?: Maybe<Array<LocalizedString>>;
  list: Scalars['Boolean'];
  name: Scalars['String'];
  nullable?: Maybe<Scalars['Boolean']>;
  readonly?: Maybe<Scalars['Boolean']>;
  type: Scalars['String'];
  ui?: Maybe<Scalars['JSON']>;
};
