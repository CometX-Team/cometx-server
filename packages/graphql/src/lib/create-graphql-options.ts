import { GqlModuleOptions, GraphQLTypesLoader } from '@nestjs/graphql';

import { buildSchema, GraphQLSchema, printSchema } from 'graphql';
import * as path from 'path';

export interface GraphQLApiOptions {
  apiType: 'shop' | 'admin';
  typePaths: string[];
  apiPath: string;
  debug: boolean;
  playground: boolean | any;
  resolverModule: any;
}

export async function createGraphQLOptions(
  typesLoader: GraphQLTypesLoader,
  options: GraphQLApiOptions,
): Promise<GqlModuleOptions> {
  const builtSchema = await buildSchemaForAPI();

  return {
    path: '/' + options.apiPath,
    typeDefs: printSchema(builtSchema),
    include: [...options.resolverModule],
    playground: options.playground || false,
    debug: options.debug || false,
    context: (req: any) => req,
    cors: false,
  } as GqlModuleOptions;

  /**
   * Generates the server's GraphQL schema by combining:
   * 1. the default schema as defined in the source .graphql files specified by `typePaths`
   * 2. any custom fields defined in the config
   * 3. any schema extensions defined by plugins
   */
  async function buildSchemaForAPI(): Promise<GraphQLSchema> {
    // Paths must be normalized to use forward-slash separators.
    // See https://github.com/nestjs/graphql/issues/336
    const normalizedPaths = options.typePaths
      .concat(['packages/graphql/src/lib/custom-fields.types.graphql'])
      .map(p => p.split(path.sep).join('/'));

    const typeDefs = await typesLoader.mergeTypesByPaths(normalizedPaths);

    const schema = buildSchema(typeDefs);

    return schema;
  }
}
