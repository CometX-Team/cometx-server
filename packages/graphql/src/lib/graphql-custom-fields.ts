import { buildSchema, extendSchema, GraphQLSchema, parse } from 'graphql';
import { getGraphQLTypes, mapToFields, wrapListType } from './utils';

/**
 * Given a CustomFields config object, generates an SDL string extending the built-in
 * types with a customFields property for all entities, translations and inputs for which
 * custom fields are defined.
 */
export function addGraphQLCustomField(
  typeDefsOrSchema: string | GraphQLSchema,
  customFieldConfig: any,
  publicOnly: boolean,
): GraphQLSchema {
  const schema =
    typeof typeDefsOrSchema === 'string'
      ? buildSchema(typeDefsOrSchema)
      : typeDefsOrSchema;

  let customFieldTypeDefs = '';

  if (!schema.getType('JSON')) {
    customFieldTypeDefs += `
      scalar JSON
    `;
  }

  if (!schema.getType('DateTime')) {
    customFieldTypeDefs += `
      scalar DateTime
    `;
  }

  for (const entityName of Object.keys(customFieldConfig)) {
    const customEntityFields = (customFieldConfig[entityName] || []).filter(
      config => {
        return (
          !config.internal &&
          (publicOnly === true ? config.public !== false : true)
        );
      },
    );

    if (schema.getType(entityName)) {
      if (customEntityFields.length) {
        customFieldTypeDefs += `
        type ${entityName}CustomFields {
            ${mapToFields(customEntityFields, wrapListType(getGraphQLTypes))}
        }

        extend type ${entityName} {
            customFields: ${entityName}CustomFields
        }
    `;
      } else {
        customFieldTypeDefs += `
          extend type ${entityName} {
            customFields: JSON
          }
        `;
      }
    }
  }

  return extendSchema(schema, parse(customFieldTypeDefs));
}
