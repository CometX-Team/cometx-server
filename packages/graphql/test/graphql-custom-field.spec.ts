import { printSchema } from 'graphql';
import { addGraphQLCustomField } from '../src/lib/graphql-custom-fields';

describe('addGraphQLCustomField()', () => {
  it('should uses JSON scalar if no custom fields defined', () => {
    const input = `
        type Product {
            id: ID
        }
    `;

    const customFieldConfig = {
      Product: [],
    };

    const result = addGraphQLCustomField(input, customFieldConfig, false);
    expect(printSchema(result)).toMatchSnapshot();
  });

  it('should extends a type', () => {
    const input = `
      type Product {
          id: ID
      }
    `;

    const customFieldConfig = {
      Product: [{ name: 'available', type: 'boolean' }],
    };

    const result = addGraphQLCustomField(input, customFieldConfig, false);

    expect(printSchema(result)).toMatchSnapshot();
  });
});
