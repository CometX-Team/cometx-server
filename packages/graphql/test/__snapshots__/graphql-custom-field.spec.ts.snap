// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addGraphQLCustomField() should extends a type 1`] = `
"type Product {
  id: ID
  customFields: ProductCustomFields
}

scalar JSON

scalar DateTime

type ProductCustomFields {
  available: Boolean
}
"
`;

exports[`addGraphQLCustomField() should uses JSON scalar if no custom fields defined 1`] = `
"type Product {
  id: ID
  customFields: JSON
}

scalar JSON

scalar DateTime
"
`;
